File: Array/904.Fruit_Into_Baskets/Python/904.Fruit_Into_Baskets.py
class Solution(object):
    def totalFruit(self, fruits):
        """
        :type fruits: List[int]
        :rtype: int
        """
        #=======================================#
        # Sliding-window based traversal method #
        #=======================================#

        ############
        #Initialize
        ##### Length of fruits array #####
        len_fruits = len(fruits)

        if (len_fruits < 3): #Issue/Boundary-case handling
            return len_fruits

        ##### Record two-pointers (i.e. left, right) #####
        record_left_ptr, record_right_ptr = 0, 0

        ##### Record dictionary/hashmap #####
        record_dict = {}

        ##### Result maximun summary counter #####
        res_max_cnter = 0


        ######################################################################
        #Sliding-window based loop traversal with recorded dictionary/hashmap
        while (record_right_ptr < len_fruits):

            ##### Check if the current indexed-value existed or not #####
            if (fruits[record_right_ptr] not in record_dict):
                record_dict[fruits[record_right_ptr]] = 1 #Keep updating/recording
            else:
                record_dict[fruits[record_right_ptr]] += 1 #Keep updating/recording

            ##### Check if the current fruits types exceed boundary or not #####
            while (len(record_dict) > 2):
                record_dict[fruits[record_left_ptr]] -= 1 #Keep updating/recording

                ##### Check if the current indexed-value matched conditions or not #####
                if (not record_dict[fruits[record_left_ptr]]):
                    del record_dict[fruits[record_left_ptr]] #Keep updating/delete
                else:
                    pass

                record_left_ptr += 1 #Keep updating/travsersing

            res_max_cnter = max(res_max_cnter, ((record_right_ptr - record_left_ptr) + 1)) #Keep updating/overwriting

            record_right_ptr += 1 #Keep updating/travsersing

        return res_max_cnter

File: Array/904.Fruit_Into_Baskets/C++/904.Fruit_Into_Baskets.cpp
class Solution {
public:
    int totalFruit(vector<int> & fruits) 
    {
        //==========================================//
        // Input type:                              //
        //   - vector<int> ref. (fruits vector)     //
        // Return type:                             //
        //   - int (result maximun summary counter) //
        //==========================================//

        /*Initialize*/
        ///// Length of fruits vector /////
        int len_fruits = fruits.size();

        if (len_fruits < 3) { return len_fruits; } //Issue, Boundary-case handling

        ///// Record dictionary, hashmap /////
        unordered_map<int, int> record_map;

        ///// Record two-pointers (i.e. left, right) /////
        int record_left_ptr = 0, record_right_ptr = 0;

        ///// Result maximun summary counter /////
        int res_max_cnter = 0;


        /*Sliding-window based loop traversal with recorded dictionary, hashmap*/
        while (record_right_ptr < len_fruits) //Whole
        {
            ///// Check if the current indexed-value existed or not /////
            if (record_map.find(fruits[record_right_ptr]) == record_map.end())
            {
                record_map[fruits[record_right_ptr]] = 1; //Keep updating, recording
            }
            else { record_map[fruits[record_right_ptr]] += 1; } //Keep updating, recording

            while (record_map.size() > 2) //Part
            {
                record_map[fruits[record_left_ptr]] -= 1; //Keep updating, recording

                ///// Check if the current indexed-value matched conditions or not ///// 
                if (! record_map[fruits[record_left_ptr]])
                {
                    record_map.erase(fruits[record_left_ptr]); //Keep updating, delete
                }
                else { ; }

                (record_left_ptr++); //Keep updating, traversing

            } //Part

            res_max_cnter = max(res_max_cnter, ((record_right_ptr - record_left_ptr) + 1)); //Keep updating, overwriting

            (record_right_ptr++); //Keep updating, traversing

        } //Whole

        return res_max_cnter;
    }
};
